saw-core/src/SAWCore/SCTypeCheck.hs

```diff
@@ -684,8 +687,18 @@ compileRecursor dt params motive cs_fs =
      let d = dtName dt
      let ctorOrder = map ctorName (dtCtors dt)
      let ctorVarIxs = map nameIndex ctorOrder
+     ixs_vars <- traverse (liftTCM scVariable) (dtIndices dt)
+     d_params <- liftTCM scConstApply d (map typedVal params)
+     d_params_ixs <- liftTCM scApplyAll d_params ixs_vars
+     x_ec <- liftTCM scFreshEC "x" d_params_ixs
+     x_var <- liftTCM scVariable x_ec
+     ty0 <- liftTCM scApplyAll (typedVal motive) (ixs_vars ++ [x_var])
+     ty1 <- liftTCM scGeneralizeExts (dtIndices dt ++ [x_ec]) ty0
+     let subst = IntMap.fromList (zip (map ecVarIndex (dtParams dt)) (map typedVal params))
+     ty2 <- liftTCM scInstantiateExt subst ty1
+     ty <- typeInferComplete ty2
```

Comment:
A lot of this code appears to be shared in common with `scRecursorRetTypeType`. Is it possible to factor out the code that they share in common?

Reply:
Good idea; I'll see if I can reuse `scRecursorRetTypeType`.